modules/cuml/src/umap.ts

```diff
@@ -0,0 +1,111 @@
+
+  private _generate_embeddings(n_samples: number): Series<Float32> {
+    return Series.sequence(
+      {type: new Float32, size: n_samples * this._umap.nComponents, init: 0, step: 0});
+  }
+
+  fit(X: Series<Numeric>|DataFrame<TypeMap>,
+      y: null|Series<Numeric>|DataFrame<TypeMap>,
+      convertDType: boolean): void {
+    const n_samples  = (X instanceof Series) ? X.length : X.numRows;
+    const n_features = (X instanceof Series) ? 1 : X.numColumns;
+    this._embeddings = transform_input_to_device_buffer(this._generate_embeddings(n_samples));
+
+    this._embeddings = this._umap.fit(transform_input_to_device_buffer(X),
+                                      n_samples,
+                                      n_features,
+                                      (y == null) ? null : transform_input_to_device_buffer(y),
+                                      null,
+                                      null,
+                                      convertDType,
+                                      this._embeddings);
+    return undefined;
+  }
+
+  _process_embeddings(embeddings: DeviceBuffer,
+                      n_samples: number,
+                      returnType: 'dataframe'|'series'|'devicebuffer') {
+    if (returnType == 'dataframe') {
+      return series_to_dataframe(
+        Series.new({type: new Float32, data: embeddings}), n_samples, this.nComponents);
+    } else if (returnType == 'series') {
+      return Series.new({type: new Float32, data: embeddings});
+    } else {
+      return embeddings;
+    }
+  }
+  fit_transform(X: Series<Numeric>|DataFrame<TypeMap>,
+                y: null|Series<Numeric>|DataFrame<TypeMap>,
+                convertDType: boolean,
+                returnType: 'dataframe'|'series'|'devicebuffer' = 'dataframe'): DataFrame
+    |Series<Float32>|DeviceBuffer {
+    const n_samples = (X instanceof Series) ? X.length : X.numRows;
+    this.fit(X, y, convertDType);
+    return this._process_embeddings(this._embeddings, n_samples, returnType);
+  }
+
+  transform(X: Series<Numeric>|DataFrame<TypeMap>,
+            convertDType: boolean,
+            returnType: 'dataframe'|'series'|'devicebuffer' = 'dataframe'): DataFrame
+    |Series<Float32>|DeviceBuffer {
+    const n_samples  = (X instanceof Series) ? X.length : X.numRows;
+    const n_features = (X instanceof Series) ? 1 : X.numColumns;
+    const embeddings = transform_input_to_device_buffer(this._generate_embeddings(n_samples));
+
+    const result = this._umap.transform(transform_input_to_device_buffer(X),
+                                        n_samples,
+                                        n_features,
+                                        null,
+                                        null,
+                                        convertDType,
+                                        this._embeddings,
+                                        embeddings);
+    return this._process_embeddings(result, n_samples, returnType);
+  }
```

Comment:
Same for these two. Split out by input type, then strongly-type the DataFrame schema. `DataFrame<TypeMap>` doesn't restrict the input columns to anything specific. It should contain numeric columns, right?

Reply:
Yes only numeric types work